client/src/components/body/profilepage/components/GroupForm.js

```diff
@@ -29,152 +27,126 @@ const GroupForm = (props) => {
+  }, [checkedLabs]);
 
-  const onFormSubmit = (e) => {
+  const toggleCheck = (labID) => {
+    setCheckedLabs((prevCheckedLabs) => ({
+      ...prevCheckedLabs,
+      [labID]: !prevCheckedLabs[labID],
+    }));
+  };
+
+  const onFormSubmit = async (e) => {
     e.preventDefault();
-    // This cursed line of code will:
-    //  1. Grab the form with e.target
-    //  2. Create an instance of a FormData object from ReactStrap using the form data
-    //  3. Get the fields in that FormData object
-    //  4. Create a new object from those entries, our new object will be an Object and not FormData
-    // Essentially, this is just a really long winded way of casting FormData to Object, since
-    // apparently that's necessary despite this not being an object-oriented or strongly typed language
-    // Please refactor this eventually, for the love of all that is holy
-    const formData = Object.fromEntries(new FormData(e.target).entries());
-    const labs = [];
-    for (const [key, value] of Object.entries(formData)) {
-      if (value === "on") {
-        labs.push(parseInt(key));
-      }
-    }
 
-    switch (addMode) {
-      case "add_instr_grp":
-        const groupName =
-          formData.groupName !== "" ? formData.groupName : "Default Group Name";
-        GroupService.createGroup(user.userid, groupName).then((data) => {
-          labs.forEach((labID) => {
-            GroupService.addGroupLab(data.groupID, labID);
-            setInstrGroupsUpdated(true);
-          });
-          setInstrGroupsUpdated(true);
+    try {
+      const formData = new FormData(e.target);
+      const groupName = formData.get("groupName") || "Default Group Name";
+
+      let selectedLabs = Object.keys(checkedLabs)
+        .filter((labID) => checkedLabs[labID])
+        .map((labID) => parseInt(labID));
+
+      if (addMode === "add_instr_grp") {
+        const newGroup = await GroupService.createGroup(user.userid, groupName);
+        console.log("New group created:", newGroup);
+
+        const addLabPromises = selectedLabs.map((labID) => {
+          GroupService.addGroupLab(newGroup.id, labID);
         });
-        break;
-      case "update_grp_lab":
-        if (groupID) {
-          if (formData.groupName !== "") {
-            GroupService.updateGroup(groupID, formData.groupName);
-          }
-          labsAssigned.forEach((labID) => {
-            if (!labs.includes(labID)) {
-              GroupService.deleteGroupLab(groupID, labID).then(() =>
-                setInstrGroupsUpdated(true),
-              );
+
+        await Promise.all(addLabPromises);
+      } else if (addMode === "update_grp_lab" && groupID) {
+        if (formData.get("groupName") !== props.groupName) {
+          await GroupService.updateGroup(groupID, formData.get("groupName"));
+        }
+
+        if (selectedLabs.length >= 0) {
+          for (const lab in checkedLabs) {
+            if (Array.isArray(assignedLabs) && assignedLabs.length !== 0) {
```

Comment:
i am a bit confused, what are we trying to do here?

Reply:
Essentially checking to see if the array, selectedLabs, is greater than 0. This was used as a check in the prior version so i kept it, but it has now been removed and it still works fine.